Disable Work Item Link Category and Type Create/Update/Delete endpoints

alexeykazakov opened the pull request:
Related to https://github.com/fabric8-services/fabric8-wit/issues/1299

fabric8cd commented:
@alexeykazakov snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1521-1`

hectorj2f commented on controller/work_item_link_category_blackbox_test.go:
```diff
@@ -142,12 +146,90 @@ func (s *workItemLinkCategorySuite) createWorkItemLinkCategoryUser() (http.Respo
+// Currently not used. Disabled as part of https://github.com/fabric8-services/fabric8-wit/issues/1299
 // TestCreateWorkItemLinkCategory tests if we can create the "test-system" work item link category
-func (s *workItemLinkCategorySuite) TestCreateAndDeleteWorkItemLinkCategory() {
+func (s *workItemLinkCategorySuite) xTestCreateAndDeleteWorkItemLinkCategory() {
```
Maybe, we could use the prefix Disabled instead of `x`. People doesn't often read the comments in the Test files :).

kwk replied:
I'd say we shouldn't use either of those but have `bool` test flag (defined in `controller_test` package) that is disabled by default. If that flag is enabled we skip the test that shall be disabled.

```go
// Define as many of such flags and skip texts somewhere
// in the controller_test package as you feel are needed.
var enableTestsLinkTypeCreation = flag.Bool("enableTestsLinkTypeCreation", false, "when set tests with link type creation are enabled")
const linkTypeCreationSkipText = "skipped because test flag \"-enableTestsLinkTypeCreation\" is not given"

func TestSomethingWithLinkTypeCreation(t *testing.T) {
  if !*enableTestsLinkTypeCreation {
    t.Skip(linkTypeCreationSkipText)
  }
}
```

I think the effort for disabling code is the same but in order to get those tests back, we have almost no work to do.

alexeykazakov replied:
It won't be enough to just pass the flag to bring the test back. They will still fail without enabled endpoints. Even if we bring our endpoints back they will need some refactoring to provide proper security. So the tests will need to be tuned too.
But I like the idea of changing function names to disableTestSomething() instead of xTestSomething().

kwk replied:
@alexeykazakov I know that those tests need fixing but we will still see in the test output that we have a test that has been skipped for a reason.

fabric8cd commented:
@alexeykazakov snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1521-2`

kwk commented on controller/work_item_children_blackbox_test.go:
```diff
@@ -154,11 +154,13 @@ func (s *workItemChildSuite) SetupTest() {
+	// _, workitemLinkCategory := test.CreateWorkItemLinkCategoryCreated(s.T(), s.svc.Context, s.svc, s.workitemLinkCategoryCtrl, createLinkCategoryPayload)
+	// require.NotNil(s.T(), workitemLinkCategory)
+	// userLinkCategoryID := *workitemLinkCategory.Data.ID
+	// s.T().Logf("Created link category with ID: %s\n", *workitemLinkCategory.Data.ID)
```
Maybe remove the commented out code?

alexeykazakov commented:
@kwk @hectorj2f thanks for review!
@kwk I removed the comments but regarding the test flag..
It won't be enough to just pass the flag to bring the tests back. They will still fail with disabled endpoints. Even if we bring our endpoints back they will need some refactoring to provide proper security. So the tests will need to be tuned too.
I changed the function names to disableTestSomething() instead of xTestSomething().
WDYT?

fabric8cd commented:
@alexeykazakov snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1521-4`

fabric8cd commented:
@alexeykazakov snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1521-5`

joshuawilson commented on controller/work_item_link_category.go:
```diff
@@ -52,12 +52,16 @@ func enrichLinkCategoryList(ctx *workItemLinkContext, list *app.WorkItemLinkCate
 // Create runs the create action.
 func (c *WorkItemLinkCategoryController) Create(ctx *app.CreateWorkItemLinkCategoryContext) error {
+	// Currently not used. Disabled as part of https://github.com/fabric8-services/fabric8-wit/issues/1299
+	if true {
```
Why does this say `if true` instead of just having the `return` statement?

alexeykazakov replied:
Because compiler will fail if there is some code after return statement.

joshuawilson replied:
ah, thanks

fabric8cd commented:
@alexeykazakov snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1521-6`

kwk commented on controller/work_item_link_category_blackbox_test.go:
```diff
@@ -142,12 +146,91 @@ func (s *workItemLinkCategorySuite) createWorkItemLinkCategoryUser() (http.Respo
+
+func createWorkItemLinkCategoryInRepo(t *testing.T, db application.DB, ctx context.Context, name string, description string, id *string) uuid.UUID {
+	var cID uuid.UUID
+	if id == nil {
```
This might be a personal preference but I hate passing `nil`. 

Please change the function signature from 
```go
func createWorkItemLinkCategoryInRepo(
t *testing.T, 
db application.DB, 
ctx context.Context, 
name string, 
description string, 
id *string) uuid.UUID {
```
to
```go
func createWorkItemLinkCategoryInRepo(
t *testing.T, 
db application.DB, 
ctx context.Context, 
linkCat link.WorkItemLinkCategory) uuid.UUID {
```
Then the caller absolutely knows what needs to be passed and if the category struct suddenly requires a new field, we don't have to modify all calls to this function.

alexeykazakov replied:
Done in https://github.com/fabric8-services/fabric8-wit/pull/1521/commits/20c8e3b1da0c48138f6a15728a67296c6a0b33bc

kwk replied:
@alexeykazakov thank you. I really do think it makes things more clear and the code seems shorter also.

kwk commented on controller/work_item_link_category_blackbox_test.go:
```diff
@@ -142,12 +146,91 @@ func (s *workItemLinkCategorySuite) createWorkItemLinkCategoryUser() (http.Respo
+	id := uuid.FromStringOrNil("bf30167a-9446-42de-82be-6b3815152051")
+
+	// Use the goa generated code to create a work item link category
+	payload := app.CreateWorkItemLinkCategoryPayload{
```
Can you instead do:

```go
appLinkCat := convertLinkCategoryFromModel(link.WorkItemLinkCategory{
  ID: id,
  Name: name,
  Description: &description,
}
payload := app.CreateWorkItemLinkCategoryPayload{Data: appLinkCat.Data}
```
That should do the conversion you're doing here manually.

alexeykazakov replied:
Done in https://github.com/fabric8-services/fabric8-wit/pull/1521/commits/20c8e3b1da0c48138f6a15728a67296c6a0b33bc

kwk commented on controller/work_item_link_category_blackbox_test.go:
```diff
@@ -142,12 +146,91 @@ func (s *workItemLinkCategorySuite) createWorkItemLinkCategoryUser() (http.Respo
+	description := "New description for work item link category."
+	id := uuid.FromStringOrNil("88727441-4a21-4b35-aabe-007f8273cd19")
+	payload := &app.UpdateWorkItemLinkCategoryPayload{
+		Data: &app.WorkItemLinkCategoryData{
```
Can you instead do:

```go
appLinkCat := convertLinkCategoryFromModel(link.WorkItemLinkCategory{
  ID: id,
  Name: name,
  Description: &description,
}
payload := app.UpdateWorkItemLinkCategoryPayload{Data: appLinkCat.Data}
```
That should do the conversion you're doing here manually.

alexeykazakov replied:
Done in https://github.com/fabric8-services/fabric8-wit/pull/1521/commits/20c8e3b1da0c48138f6a15728a67296c6a0b33bc

fabric8cd commented:
@alexeykazakov snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1521-7`

alexeykazakov commented:
Yeah.. these tests fail from time to time for some reason. Not related to this PR though.
Thanks for the review!
